bdai_ros2_wrappers/bdai_ros2_wrappers/logging.py

```diff
@@ -37,6 +44,229 @@
         warnings.warn("TODO: use subloggers in RcutilsLogHandler implementation", stacklevel=1)
 
 
+@functools.lru_cache(maxsize=1024)
+def make_logging_function(
+    name: str,
+    level: LoggingSeverity,
+    origin: inspect.Traceback,
+    throttle_duration_sec: typing.Optional[float] = None,
+    throttle_time_source: typing.Optional[Clock] = None,
+    skip_first: typing.Optional[bool] = None,
+    once: typing.Optional[bool] = None,
+) -> typing.Callable[[str], bool]:
+    """Make a fast rclpy logging function.
+
+    A logging function takes a log message and returns True if the message was logged, False otherwise.
+    A message will not be logged if:
+
+    * the logger is not enabled for the logging function's severity level (i.e. it is less than
+      the severity level of the logger), or
+    * some logging filter (throttling, once, skip) causes the message to be skipped.
+
+    Logging filters will only be evaluated if the logger is enabled for the logging function's severity level.
+
+    Args:
+        name: logger name.
+        level: log severity level.
+        origin: log call site.
+        throttle_duration_sec: optional period for throttling, in seconds.
+        throttle_time_source: optional time source for throttling, defaults to system time.
+        skip_first: if True, skip the first log.
+        once: if True, only log once.
+
+    Returns:
+        a logging function.
+    """
+
+    def log(message: str) -> bool:
+        impl.rclpy_logging_rcutils_log(
+            level,
+            name,
+            message,
+            origin.function,
+            origin.filename,
+            origin.lineno,
+        )
+        return True
+
+    if throttle_duration_sec is not None:
+        throttle_period = Duration(
+            seconds=throttle_duration_sec,
+        )
+        log = throttle(
+            log,
+            min_period=throttle_period,
+            time_source=throttle_time_source,
+            fill_value=False,
+        )
+    if skip_first:
+        log = skip(log, num_times=1, fill_value=False)
+    if once:
+        log = cap(log, num_times=1, fill_value=False)
+    return log
+
+
+class MemoizingRcutilsLogger:
```

Comment:
This this class name a standard one? The name (`memoizing` sounds a bit new to me.

Reply:
Hmm, not standard, no. Memoizing is me adjectivizing [memoization](https://en.wikipedia.org/wiki/Memoization).